File: src/windy/debugger.py
"""Interactive step debugger for Windy.

Renders a viewport of the grid (with the IP highlighted), the stack, the
run-time state, and captured program output, pausing for command input
between steps.

Commands:

  Enter / s / step    advance one step
  c / cont / run      run until HALT or the built-in safety cap
  q / quit            exit the debugger

``debug_source`` is the library entry point. It accepts an ``input_fn``
hook so tests can drive it without a tty.
"""

from __future__ import annotations

import sys
from collections.abc import Callable
from io import StringIO
from typing import IO

from rich.console import Console, Group
from rich.panel import Panel
from rich.text import Text

from .compiler import CompiledGrid
from .easter import BANNER, detect
from .ir import IP
from .opcodes import Op
from .parser import parse
from .vm import VM

VIEWPORT_COLS = 60
VIEWPORT_ROWS = 13
RUN_SAFETY_CAP = 10_000_000

DIR_NAMES: dict[tuple[int, int], str] = {
    (1, 0): "→ east",
    (1, -1): "↗ ne",
    (0, -1): "↑ north",
    (-1, -1): "↖ nw",
    (-1, 0): "← west",
    (-1, 1): "↙ sw",
    (0, 1): "↓ south",
    (1, 1): "↘ se",
}


def _render_grid(cgrid: CompiledGrid, ip: IP) -> Text:
    """Viewport of the grid centered on the IP, with the IP cell highlighted."""
    x0 = ip.x - VIEWPORT_COLS // 2
    y0 = ip.y - VIEWPORT_ROWS // 2
    text = Text()
    for y in range(y0, y0 + VIEWPORT_ROWS):
        for x in range(x0, x0 + VIEWPORT_COLS):
            cp = cgrid.cell(x, y)
            try:
                ch = chr(cp)
            except (ValueError, OverflowError):
                ch = "?"
            if ch in ("\n", "\r", "\t") or not ch.isprintable():
                ch = " "
            if x == ip.x and y == ip.y:
                text.append(ch, style="bold reverse magenta")
            else:
                text.append(ch)
        text.append("\n")
    return text


def _render_stack(stack: list[int]) -> Text:
    if not stack:
        return Text("(empty)", style="dim")
    text = Text()
    for v in reversed(stack[-20:]):
        text.append(f"{v}\n")
    if len(stack) > 20:
        text.append(f"... (+{len(stack) - 20} below)\n", style="dim")
    return text


def _render_status(vm: VM) -> Text:
    text = Text()
    text.append(f"step:   {vm.steps}\n")
    text.append(f"ip:     ({vm.ip.x}, {vm.ip.y})\n")
    text.append(f"dir:    {DIR_NAMES.get((vm.ip.dx, vm.ip.dy), '?')}\n")
    text.append(f"strmod: {'on' if vm.strmode else 'off'}\n")
    text.append(f"halted: {'yes' if vm.halted else 'no'}\n")
    cp = vm.cgrid.cell(vm.ip.x, vm.ip.y)
    try:
        ch = chr(cp)
    except (ValueError, OverflowError):
        ch = "?"
    op, operand = vm.cgrid.op(vm.ip.x, vm.ip.y)
    text.append(f"cell:   {ch!r} (U+{cp:04X})\n")
    text.append(f"op:     {op.name}")
    if op is Op.PUSH_DIGIT:
        text.append(f"  (value={operand})")
    return text


def _render_frame(console: Console, vm: VM, output: str) -> None:
    grid_panel = Panel(_render_grid(vm.cgrid, vm.ip), title="grid", border_style="blue")
    status_panel = Panel(_render_status(vm), title="state", border_style="yellow")
    stack_panel = Panel(_render_stack(vm.stack), title="stack", border_style="cyan")
    out_panel = Panel(Text(output or "(no output)"), title="stdout", border_style="green")
    console.print(grid_panel)
    console.print(Group(status_panel, stack_panel))
    console.print(out_panel)


def debug_source(
    source: str,
    *,
    stdin: IO[str] | None = None,
    console: Console | None = None,
    input_fn: Callable[[], str] | None = None,
) -> int:
    """Step-debug a Windy program. Returns the final VM exit code.

    ``input_fn`` lets callers (and tests) feed commands without a tty.
    """
    grid, scan_text = parse(source)
    console = console or Console()
    if detect(scan_text):
        print(BANNER, file=sys.stderr)

    output_buf = StringIO()
    vm = VM(
        grid,
        stdin=stdin if stdin is not None else sys.stdin,
        stdout=output_buf,
        stderr=sys.stderr,
    )

    prompt = input_fn or (lambda: console.input("[bold]> [/]"))
    console.print(
        Text(
            "windy debugger — [Enter]/s: step,  c: run-to-halt,  q: quit",
            style="dim",
        )
    )

    while True:
        _render_frame(console, vm, output_buf.getvalue())
        if vm.halted:
            console.print(Text("program halted", style="bold green"))
            return 0
        try:
            cmd = prompt().strip().lower()
        except (EOFError, KeyboardInterrupt):
            return 0

        if cmd in ("", "s", "step"):
            vm.step()
            vm.steps += 1
        elif cmd in ("c", "cont", "continue", "r", "run"):
            remaining = RUN_SAFETY_CAP
            while not vm.halted and remaining > 0:
                vm.step()
                vm.steps += 1
                remaining -= 1
            _render_frame(console, vm, output_buf.getvalue())
            if not vm.halted:
                console.print(Text("run budget exhausted", style="bold red"))
                return 124
            console.print(Text("program halted", style="bold green"))
            return 0
        elif cmd in ("q", "quit", "exit"):
            return 0
        else:
            console.print(Text(f"unknown command: {cmd!r}", style="red"))

File: src/debugger.rs
//! Interactive step debugger — terminal edition.
//!
//! No TUI crate: just ANSI escapes + Unicode box-drawing. Between steps
//! the screen is redrawn with a grid viewport around the IP, the stack,
//! runtime state, and the program's captured stdout. The user drives
//! the stepper with single-character commands:
//!
//! * `Enter` / `s` — advance one step
//! * `c` — run until HALT or the built-in safety cap
//! * `q` — quit
//!
//! `debug_source` owns the captured stdout buffer across steps so it
//! can be displayed between instructions. This is why `Vm::step`
//! accepts streams as parameters rather than owning them — see the
//! module comment in `vm.rs`.

use crate::easter::{detect, BANNER};
use crate::opcodes::{decode_cell, Op};
use crate::parser::parse;
use crate::vm::Vm;
use num_traits::ToPrimitive;
use std::io::{self, BufRead, Write};

const VIEWPORT_COLS: i64 = 60;
const VIEWPORT_ROWS: i64 = 13;
const RUN_SAFETY_CAP: u64 = 10_000_000;

const CLEAR_SCREEN: &str = "\x1b[2J\x1b[H";
const REVERSE_ON: &str = "\x1b[7m";
const REVERSE_OFF: &str = "\x1b[0m";

fn dir_name(dx: i64, dy: i64) -> &'static str {
    match (dx, dy) {
        (1, 0) => "→ east",
        (1, -1) => "↗ ne",
        (0, -1) => "↑ north",
        (-1, -1) => "↖ nw",
        (-1, 0) => "← west",
        (-1, 1) => "↙ sw",
        (0, 1) => "↓ south",
        (1, 1) => "↘ se",
        _ => "?",
    }
}

fn render_grid(out: &mut dyn Write, vm: &Vm) -> io::Result<()> {
    let half_w = VIEWPORT_COLS / 2;
    let half_h = VIEWPORT_ROWS / 2;
    let x0 = vm.ip.x - half_w;
    let y0 = vm.ip.y - half_h;
    writeln!(out, "┌─ grid {:>52}", format!("({}, {})  ", vm.ip.x, vm.ip.y))?;
    for y in y0..y0 + VIEWPORT_ROWS {
        write!(out, "│ ")?;
        for x in x0..x0 + VIEWPORT_COLS {
            let cp = vm.grid.get(x, y);
            let ch = cp
                .to_u32()
                .and_then(char::from_u32)
                .filter(|c| !c.is_control() && *c != '\t' && *c != '\n')
                .unwrap_or(' ');
            if x == vm.ip.x && y == vm.ip.y {
                write!(out, "{REVERSE_ON}{ch}{REVERSE_OFF}")?;
            } else {
                write!(out, "{ch}")?;
            }
        }
        writeln!(out, " │")?;
    }
    writeln!(out, "└{}", "─".repeat(VIEWPORT_COLS as usize + 2))?;
    Ok(())
}

fn render_status(out: &mut dyn Write, vm: &Vm) -> io::Result<()> {
    let cell = vm.grid.get(vm.ip.x, vm.ip.y);
    let cp = cell.to_u32().unwrap_or(0);
    let ch = char::from_u32(cp).unwrap_or('?');
    let (op, operand) = decode_cell(&cell);
    writeln!(out, "┌─ state ─────")?;
    writeln!(out, "│ step:   {}", vm.steps)?;
    writeln!(out, "│ ip:     ({}, {})", vm.ip.x, vm.ip.y)?;
    writeln!(out, "│ dir:    {}", dir_name(vm.ip.dx, vm.ip.dy))?;
    writeln!(out, "│ strmod: {}", if vm.strmode { "on" } else { "off" })?;
    writeln!(out, "│ halted: {}", if vm.halted { "yes" } else { "no" })?;
    writeln!(out, "│ cell:   {:?} (U+{:04X})", ch, cp)?;
    let op_line = if op == Op::PushDigit {
        format!("│ op:     {} (value={})", op.name(), operand)
    } else {
        format!("│ op:     {}", op.name())
    };
    writeln!(out, "{op_line}")?;
    writeln!(out, "└─────────────")?;
    Ok(())
}

fn render_stack(out: &mut dyn Write, vm: &Vm) -> io::Result<()> {
    writeln!(out, "┌─ stack (top first) ─")?;
    if vm.stack.is_empty() {
        writeln!(out, "│ (empty)")?;
    } else {
        let show = 20usize.min(vm.stack.len());
        for v in vm.stack.iter().rev().take(show) {
            writeln!(out, "│ {v}")?;
        }
        if vm.stack.len() > show {
            writeln!(out, "│ ... (+{} below)", vm.stack.len() - show)?;
        }
    }
    writeln!(out, "└─────────────────────")?;
    Ok(())
}

fn render_stdout(out: &mut dyn Write, captured: &[u8]) -> io::Result<()> {
    writeln!(out, "┌─ program stdout ─────")?;
    if captured.is_empty() {
        writeln!(out, "│ (no output)")?;
    } else {
        let text = String::from_utf8_lossy(captured);
        for line in text.lines() {
            writeln!(out, "│ {line}")?;
        }
        if !text.ends_with('\n') && !text.is_empty() {
            // No trailing newline — the last line has already been emitted,
            // so just close the panel.
        }
    }
    writeln!(out, "└──────────────────────")?;
    Ok(())
}

fn render_frame(out: &mut dyn Write, vm: &Vm, captured: &[u8]) -> io::Result<()> {
    write!(out, "{CLEAR_SCREEN}")?;
    writeln!(
        out,
        "windy debugger — [Enter]/s: step  c: run-to-halt  q: quit"
    )?;
    render_grid(out, vm)?;
    render_status(out, vm)?;
    render_stack(out, vm)?;
    render_stdout(out, captured)?;
    out.flush()?;
    Ok(())
}

/// Run the interactive debugger over `source`, returning the final exit code.
///
/// Reads commands from `stdin_commands`. The VM's own stdin is fixed as
/// an empty byte slice — Windy programs under the debugger can't prompt
/// for input because the real stdin is owned by the debugger loop.
pub fn debug_source(source: &str, stdin_commands: &mut dyn BufRead) -> i32 {
    debug_source_inner(source, stdin_commands, &mut io::stderr(), &mut io::stdout())
}

pub(crate) fn debug_source_inner(
    source: &str,
    commands: &mut dyn BufRead,
    stderr: &mut dyn Write,
    ui: &mut dyn Write,
) -> i32 {
    let (grid, scan_text) = parse(source);
    if detect(&scan_text) {
        let _ = writeln!(stderr, "{}", BANNER);
    }

    let mut captured_out: Vec<u8> = Vec::new();
    let mut captured_err: Vec<u8> = Vec::new();
    let mut empty_stdin: &[u8] = b"";
    let mut vm = Vm::new(grid, None, None);

    loop {
        let _ = render_frame(ui, &vm, &captured_out);
        if vm.halted {
            let _ = writeln!(ui, "program halted");
            return 0;
        }
        let _ = write!(ui, "> ");
        let _ = ui.flush();

        let mut line = String::new();
        match commands.read_line(&mut line) {
            Ok(0) | Err(_) => return 0,
            _ => {}
        }
        let cmd = line.trim().to_lowercase();

        match cmd.as_str() {
            "" | "s" | "step" => {
                vm.step(&mut empty_stdin, &mut captured_out, &mut captured_err);
                vm.steps += 1;
            }
            "c" | "cont" | "continue" | "r" | "run" => {
                let mut remaining = RUN_SAFETY_CAP;
                while !vm.halted && remaining > 0 {
                    vm.step(&mut empty_stdin, &mut captured_out, &mut captured_err);
                    vm.steps += 1;
                    remaining -= 1;
                }
                let _ = render_frame(ui, &vm, &captured_out);
                if !vm.halted {
                    let _ = writeln!(ui, "run budget exhausted");
                    return 124;
                }
                let _ = writeln!(ui, "program halted");
                return 0;
            }
            "q" | "quit" | "exit" => return 0,
            _ => {
                let _ = writeln!(ui, "unknown command: {:?}", cmd);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    fn drive(source: &str, commands: &str) -> (i32, String, String) {
        let mut cmds = Cursor::new(commands.as_bytes());
        let mut stderr: Vec<u8> = Vec::new();
        let mut ui: Vec<u8> = Vec::new();
        let code = debug_source_inner(source, &mut cmds, &mut stderr, &mut ui);
        (
            code,
            String::from_utf8(ui).unwrap(),
            String::from_utf8(stderr).unwrap(),
        )
    }

    #[test]
    fn quit_returns_zero() {
        let (code, _, _) = drive("@", "q\n");
        assert_eq!(code, 0);
    }

    #[test]
    fn step_then_quit() {
        let (code, _, _) = drive("5.@", "s\nq\n");
        assert_eq!(code, 0);
    }

    #[test]
    fn continue_runs_to_halt() {
        let (code, ui, _) = drive("\"A\",@", "c\n");
        assert_eq!(code, 0);
        assert!(ui.contains("program halted"));
        // Captured stdout of the inner program appears in the UI buffer.
        assert!(ui.contains("A"));
    }

    #[test]
    fn halt_detected_after_single_step() {
        let (code, _, _) = drive("@", "s\n");
        assert_eq!(code, 0);
    }

    #[test]
    fn unknown_command_is_surfaced_not_fatal() {
        let (code, ui, _) = drive("@", "nope\nc\n");
        assert_eq!(code, 0);
        assert!(ui.contains("unknown command"));
    }

    #[test]
    fn sisobus_banner_goes_to_stderr() {
        let (_, ui, err) = drive("\"sisobus\"@", "q\n");
        assert!(err.contains("Kim Sangkeun"));
        assert!(!ui.contains("Kim Sangkeun"));
    }
}
